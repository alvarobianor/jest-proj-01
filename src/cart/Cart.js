import { find, findIndex, remove } from 'lodash';

export default class Cart {
  items = [];

  getTotal() {
    return this.items.reduce((old, item) => {
      return old + item.product.price * item.quantity;
    }, 0);
  }

  add(item) {
    const itemToFind = { product: item.product };
    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }
    this.items.push(item);
  }

  add2(item) {
    const itemToFind = { product: item.product };
    const indexItem = findIndex(this.items, itemToFind);
    if (indexItem > -1) {
      this.items[indexItem].quantity += item.quantity;
    } else {
      this.items.push(item);
    }
  }

  remove(product) {
    remove(this.items, { product });
  }

  checkout() {
    return {
      total: this.getTotal(),
      items: this.items,
    };
  }
}
