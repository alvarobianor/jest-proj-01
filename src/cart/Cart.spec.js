import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let product;
  let product2;

  beforeEach(() => {
    cart = new Cart();
    product = {
      title: 'Adidas running shoes',
      price: 14990,
    };

    product2 = {
      title: 'moletom Air Jodan',
      price: 32990,
    };
  });

  describe('getTotal()', () => {
    it('return 0 when the method getTotal() is called in a newly created instance', () => {
      expect(cart.getTotal()).toEqual(0);
    });

    it('should multiply quantity and price, and receive the total amount', () => {
      cart.add({
        product,
        quantity: 4, // 59.960
      });

      expect(cart.getTotal()).toEqual(59960);
    });

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 1,
      });

      cart.add({
        product,
        quantity: 4, // 59.960
      });

      expect(cart.getTotal()).toEqual(59960);
    });

    it('should update the quantity of a product that exists on cart when added the same product', () => {
      cart.add2({
        product,
        quantity: 2,
      });

      cart.add2({
        product,
        quantity: 4, // 59.960
      });

      expect(cart.getTotal()).toEqual(89940);
    });

    it('should update the total when a product gets include or removed', () => {
      cart.add({
        product,
        quantity: 2, //29.980
      });

      cart.add({
        product: product2,
        quantity: 4, // 131.960
      });

      cart.remove(product);

      expect(cart.getTotal()).toEqual(131960);
    });
  });

  describe('checkout', () => {
    it('should return an object with the total and the list of itens of cart', () => {
      cart.add({
        product,
        quantity: 54,
      });

      cart.add({
        product: product2,
        quantity: 48,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should reset the cart when the checkout method is called', () => {
      cart.add({
        product: product2,
        quantity: 48,
      });

      cart.checkout();

      expect(cart.getTotal()).toEqual(0);
    });
  });

  it('should return an object with the total and the list of itens of cart, when the summary is called', () => {
    cart.add({
      product,
      quantity: 2,
    });

    cart.add({
      product: product2,
      quantity: 4,
    });

    expect(cart.summary()).toMatchSnapshot();
    expect(cart.getTotal()).toBeGreaterThan(0);
  });
});
