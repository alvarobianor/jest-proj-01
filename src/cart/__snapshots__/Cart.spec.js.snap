// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should return an object with the total and the list of itens of cart 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 14990,
        "title": "Adidas running shoes",
      },
      "quantity": 54,
    },
    Object {
      "product": Object {
        "price": 32990,
        "title": "moletom Air Jodan",
      },
      "quantity": 48,
    },
  ],
  "total": 2392980,
}
`;

exports[`Cart should return an object with the total and the list of itens of cart, when the summary is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 14990,
        "title": "Adidas running shoes",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 32990,
        "title": "moletom Air Jodan",
      },
      "quantity": 4,
    },
  ],
  "total": 161940,
}
`;
